README.txt
Tesseract Data Provider based on tags from extension "tagpack".

Please refer to the manual for more details.


ChangeLog
2012-07-10 Francois Suter (Cobweb)  <typo3@cobweb.ch>

	* Released to TER as version 1.2.0

2012-06-29 Francois Suter (Cobweb)  <typo3@cobweb.ch>

	* Added definition of search fields, resolves #38486

2012-06-26 Francois Suter (Cobweb)  <typo3@cobweb.ch>

	* Cleaned up some code, verified compatibility with TYPO3 4.7, raised requirements to TYPO3 4.5, resolves #38401

2012-04-19 Francois Suter (Cobweb)  <typo3@cobweb.ch>

	* Released to TER as version 1.1.1

2012-04-03 Francois Suter (Cobweb)  <typo3@cobweb.ch>

	* Implemented new workflow for empty data structure creation, resolves #35053

2010-12-28 Francois Suter (Cobweb)  <typo3@cobweb.ch>

	* Released to TER as version 1.1.0

2010-12-21 Francois Suter (Cobweb)  <typo3@cobweb.ch>

	* Added support for arrays as expression results, resolves #10068
	* Added warning (in manual and code) about Data Filters not being supported yet, resolves #11231

2009-09-27 Francois Suter (Cobweb)  <typo3@cobweb.ch>

	* First public release

tca.php
<?php
if (!defined ('TYPO3_MODE')) 	die ('Access denied.');

$TCA['tx_tagpackprovider_selections'] = array (
	'ctrl' => $TCA['tx_tagpackprovider_selections']['ctrl'],
	'interface' => array (
		'showRecordFieldList' => 'name,tables,tags'
	),
	'feInterface' => $TCA['tx_tagpackprovider_selections']['feInterface'],
	'columns' => array (
		'name' => array (		
			'exclude' => 0,		
			'label' => 'LLL:EXT:tagpackprovider/locallang_db.xml:tx_tagpackprovider_selections.name',		
			'config' => array (
				'type' => 'input',
				'size' => '30',
				'eval' => 'required,trim',
			)
		),
		'tables' => array (		
			'exclude' => 0,		
			'label' => 'LLL:EXT:tagpackprovider/locallang_db.xml:tx_tagpackprovider_selections.tables',		
			'config' => array (
				'type' => 'select',
				'itemsProcFunc' => 'tx_tagpackprovider_tca->getListOfTables',
				'size' => 5,
				'maxitems' => 10,
			)
		),
		'tags' => array (		
			'exclude' => 0,		
			'label' => 'LLL:EXT:tagpackprovider/locallang_db.xml:tx_tagpackprovider_selections.tags',		
			'config' => array (
				'type' => 'group',
				'internal_type' => 'db',
				'allowed' => 'tx_tagpack_tags',
				'size' => 5,
				'minitems' => 0,
				'maxitems' => 1000,
			)
		),
		'tag_expressions' => array (
			'exclude' => 0,
			'label' => 'LLL:EXT:tagpackprovider/locallang_db.xml:tx_tagpackprovider_selections.tag_expressions',
			'config' => array (
				'type' => 'text',
				'cols' => 50,
				'rows' => 5,
			)
		),
		'logical_operator' => array(
			'exclude' => 0,
			'label' => 'LLL:EXT:tagpackprovider/locallang_db.xml:tx_tagpackprovider_selections.logical_operator',
			'config' => array (
				'type' => 'radio',
				'default' => 'OR',
				'items' => array (
					array('LLL:EXT:tagpackprovider/locallang_db.xml:tx_tagpackprovider_selections.logical_operator.I.0', 'AND'),
					array('LLL:EXT:tagpackprovider/locallang_db.xml:tx_tagpackprovider_selections.logical_operator.I.1', 'OR'),
				),
			)
		),
		'tags_override' => array (
			'exclude' => 0,
			'label' => 'LLL:EXT:tagpackprovider/locallang_db.xml:tx_tagpackprovider_selections.tags_override',
			'config' => array (
				'type' => 'check',
				'default' => 1
			)
		),
	),
	'types' => array (
		'0' => array('showitem' => 'name;;;;1-1-1, tables;;;;2-2-2, tags;;;;3-3-3, tag_expressions;;1, logical_operator')
	),
	'palettes' => array (
		'1' => array('showitem' => 'tags_override')
	)
);
?>

ext_tables.php
<?php
if (!defined ('TYPO3_MODE')) {
	die ('Access denied.');
}

t3lib_extMgm::allowTableOnStandardPages('tx_tagpackprovider_selections');


if (TYPO3_MODE == 'BE')	{
	include_once(t3lib_extMgm::extPath('tagpackprovider') . 'class.tx_tagpackprovider_tca.php');
}

$TCA['tx_tagpackprovider_selections'] = array (
	'ctrl' => array (
		'title'     => 'LLL:EXT:tagpackprovider/locallang_db.xml:tx_tagpackprovider_selections',
		'label'     => 'name',
		'tstamp'    => 'tstamp',
		'crdate'    => 'crdate',
		'cruser_id' => 'cruser_id',
		'default_sortby' => 'ORDER BY crdate',
		'delete' => 'deleted',
		'searchFields' => 'name,tables,tags,tag_expressions',
		'dynamicConfigFile' => t3lib_extMgm::extPath($_EXTKEY) . 'tca.php',
		'iconfile'          => t3lib_extMgm::extRelPath($_EXTKEY) . 'res/icon_tx_tagpackprovider_selections.gif',
	),
);

	// Add context sensitive help (csh) for this table
t3lib_extMgm::addLLrefForTCAdescr('tx_tagpackprovider_selections', t3lib_extMgm::extPath($_EXTKEY) . 'locallang_csh_txtagpackproviderselections.xml');

	// Register tagpackprovider as a secondary Data Provider
t3lib_div::loadTCA('tt_content');
$TCA['tt_content']['columns']['tx_displaycontroller_provider2']['config']['allowed'] .= ',tx_tagpackprovider_selections';

	// Add a wizard for adding a tagpackprovider
$addTagpackProviderWizard = array(
	'type' => 'script',
	'title' => 'LLL:EXT:tagpackprovider/locallang_db.xml:wizards.add_tagpackprovider',
	'script' => 'wizard_add.php',
	'icon' => 'EXT:tagpackprovider/res/add_tagpackprovider_wizard.gif',
	'params' => array(
			'table' => 'tx_tagpackprovider_selections',
			'pid' => '###CURRENT_PID###',
			'setValue' => 'append'
		)
	);
$TCA['tt_content']['columns']['tx_displaycontroller_provider2']['config']['wizards']['add_tagpackprovider'] = $addTagpackProviderWizard;
?>

class.tx_tagpackprovider.php
<?php
/***************************************************************
*  Copyright notice
*
*  (c) 2009-2010 Francois Suter (Cobweb) <typo3@cobweb.ch>
*  All rights reserved
*
*  This script is part of the TYPO3 project. The TYPO3 project is
*  free software; you can redistribute it and/or modify
*  it under the terms of the GNU General Public License as published by
*  the Free Software Foundation; either version 2 of the License, or
*  (at your option) any later version.
*
*  The GNU General Public License can be found at
*  http://www.gnu.org/copyleft/gpl.html.
*
*  This script is distributed in the hope that it will be useful,
*  but WITHOUT ANY WARRANTY; without even the implied warranty of
*  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
*  GNU General Public License for more details.
*
*  This copyright notice MUST APPEAR in all copies of the script!
***************************************************************/

/**
 * Base dataprovider service. All Data Provider services should inherit from this class
 *
 * @author		Francois Suter (Cobweb) <typo3@cobweb.ch>
 * @package		TYPO3
 * @subpackage	tx_tagpackprovider
 *
 * $Id$
 */
class tx_tagpackprovider extends tx_tesseract_providerbase {

	/**
	 * This method returns the type of data structure that the Data Provider can prepare
	 *
	 * @return	string	type of the provided data structure
	 */
	public function getProvidedDataStructure() {
		return tx_tesseract::IDLIST_STRUCTURE_TYPE;
	}

	/**
	 * This method indicates whether the Data Provider can create the type of data structure requested or not
	 *
	 * @param	string		$type: type of data structure
	 * @return	boolean		true if it can handle the requested type, false otherwise
	 */
	public function providesDataStructure($type) {
		return tx_tesseract::IDLIST_STRUCTURE_TYPE;
	}

	/**
	 * This method returns the type of data structure that the Data Provider can receive as input
	 *
	 * @return	string	type of used data structures
	 */
	public function getAcceptedDataStructure() {
		return '';
	}

	/**
	 * This method indicates whether the Data Provider can use as input the type of data structure requested or not
	 *
	 * @param	string		$type: type of data structure
	 * @return	boolean		true if it can use the requested type, false otherwise
	 */
	public function acceptsDataStructure($type) {
		return FALSE;
	}

	/**
	 * Assembles the data structure and returns it
	 *
	 * @return array Standardized data structure
	 */
	public function getDataStructure() {
			// Get the list of tables
			// The first one is considered the "main" one
		$tables = t3lib_div::trimExplode(',', $this->providerData['tables']);
		$mainTable = $tables[0];
			// If an empty structure should be returned, initialize one
		if ($this->hasEmptyOutputStructure) {
			$structure = $this->initEmptyDataStructure($mainTable, tx_tesseract::IDLIST_STRUCTURE_TYPE);

			// Otherwise, assemble the structure
		} else {
			$structure = array();
			$uidsPerTable = array();
			$tagsPerItem = array();

				// Assemble list of tags
			$manualTags = array();
				// Get manually selected tags
			if (!empty($this->providerData['tags'])) {
				$manualTags = t3lib_div::trimExplode(',', $this->providerData['tags']);
			}
				// Get tags from expressions
			$expressionTags = array();
			if (!empty($this->providerData['tag_expressions'])) {
				$expressionTags = $this->parseExpressionField($this->providerData['tag_expressions']);
			}
				// Assemble final tags array
			$tags = $manualTags;
			if (count($expressionTags) > 0) {
					// Expression tags override manual tags
				if ($this->providerData['tags_override']) {
					$tags = $expressionTags;

					// Expression tags and manual tags are merged
				} else {
					$tags = array_merge($manualTags, $expressionTags);
				}
			}
				// Make sure each tag appears only once
			$finalTags = array_unique($tags);

			$where = '';
				// Assemble where clause based on selected tags, if any
			if (count($finalTags) > 0) {
				$where = 'uid_local IN (' . implode(',', $finalTags) . ')';
			}

				// Continue only if where clause is not empty
				// (the contrary would mean selecting all records from the selected tables, which makes no sense)
			if (!empty($where)) {
					// Add condition on tables
				if (!empty($this->providerData['tables'])) {
					$condition = '';
					foreach ($tables as $aTable) {
						if (!empty($condition)) {
							$condition .= ',';
						}
						$condition .= "'" . $aTable . "'";
					}
					$where .= ' AND tablenames IN (' . $condition . ')';
				}
					// NOTE: not sure how tagpack uses/sets these fields
					// Anyway there's no TCA for table tx_tagpack_tags_relations_mm,
					// so the proper API cannot be used
				$where .= " AND hidden='0' AND deleted='0'";
					// Query the tags relations table
					// NOTE: results are sorted by table and by sorting
					// This respects the order in which tags were applied. Maybe this doesn't make sense after all. Could be reviewed at a later stage
				$res = $GLOBALS['TYPO3_DB']->exec_SELECTquery('uid_local, uid_foreign, tablenames', 'tx_tagpack_tags_relations_mm', $where, '', 'tablenames ASC, sorting ASC');

					// Loop on the results and sort them by table
				while ($row = $GLOBALS['TYPO3_DB']->sql_fetch_assoc($res)) {
					if (!isset($uidsPerTable[$row['tablenames']])) {
						$uidsPerTable[$row['tablenames']] = array();
						$tagsPerItem[$row['tablenames']] = array();
					}
					if (!isset($tagsPerItem[$row['tablenames']][$row['uid_foreign']])) {
						$tagsPerItem[$row['tablenames']][$row['uid_foreign']] = array();
					}
					$uidsPerTable[$row['tablenames']][$row['uid_foreign']] = $row['uid_foreign'];
					$tagsPerItem[$row['tablenames']][$row['uid_foreign']][] = $row['uid_local'];
				}
					// If the items should match all tags ("AND" logical operator chosen)
					// perform some post-process filtering, because such a condition
					// cannot be expressed simply in the SELECT query
				if ($this->providerData['logical_operator'] == 'AND') {
					foreach ($tagsPerItem as $table => $tableRows) {
						foreach ($tableRows as $uid_foreign => $uid_locals) {
								// Check if all chosen tags are matched by tags found per item
							$rest = array_diff($finalTags, $uid_locals);
								// At least one tag was not matched,
								// remove item from list
							if (count($rest) > 0) {
								unset($uidsPerTable[$table][$uid_foreign]);
							}
						}
					}
				}
			}

				// Assemble data structure parts
			$count = 0;
			$uniqueTable = $mainTable;
			$uidList = '';
			$uidListWithTable = '';
			if (count($uidsPerTable) > 0) {
					// Set unique table name, if indeed unique
				if (count($uidsPerTable) == 1) {
					$uniqueTable = key($uidsPerTable);
					$Uids = array_unique($uidsPerTable[$uniqueTable]);
					$uidList = implode(',', $Uids);
				} else {
					$uniqueTable = '';
					$uidList = '';
				}
					// Loop on list of uid's per table and assemble lists of id's prepended with table names
				$prependedUids = array();
				$count = 0;
				foreach ($uidsPerTable as $aTable => $list) {
					$Uids = array_unique($list);
					foreach ($Uids as $id) {
						$prependedUids[] = $aTable . '_' . $id;
						$count++;
					}
				}
				$uidListWithTable = implode(',', $prependedUids);
			}

				// Assemble final structure
			$structure['uniqueTable'] = $uniqueTable;
			$structure['uidList'] = $uidList;
			$structure['uidListWithTable'] = $uidListWithTable;
			$structure['count'] = $count;
			$structure['totalCount'] = $count;
		}
		return $structure;
	}

	/**
	 * This method is used to pass a data structure to the Data Provider
	 *
	 * @param 	array	$structure: standardised data structure
	 * @return	void
	 */
	public function setDataStructure($structure) {
	}

	/**
	 * This method is used to pass a Data Filter structure to the Data Provider
	 *
	 * @param	array	$filter: Data Filter structure
	 * @return	void
	 */
	public function setDataFilter($filter) {
			// TODO: improve logging when we finally have a central Tesseract debugging workflow
		if (TYPO3_DLOG) {
			t3lib_div::devLog('Data filters are currently not supported!', 'tagpackprovider', 2);
		}
	}

	/**
     * This method returns a list of tables and fields available in the data structure,
     * complete with localized labels
     *
     * @param	string	$language: 2-letter iso code for language
     * @return	array	list of tables and fields
     */
	public function getTablesAndFields($language = '') {
		return array();
	}

	/**
	 * This method relies on the expressions parser to get tags from expressions
	 *
	 * @param	string	$field: field to parse for expressions
	 * @return	string	Comma-separated list of tag primary keys
	 */
	protected function parseExpressionField($field) {
		$tags = array();
			// Parse the field
		$allLines = tx_tesseract_utilities::parseConfigurationField($field);
			// Interpret each line
		foreach ($allLines as $line) {
			try {
				$evaluatedExpression = tx_expressions_parser::evaluateExpression($line);
				if (!empty($evaluatedExpression)) {
					if (is_array($evaluatedExpression)) {
						$tagList = $evaluatedExpression;
					} elseif (strpos($evaluatedExpression, ',') === false) {
						$tagList = array($evaluatedExpression);
					} else {
						$tagList = t3lib_div::trimExplode(',', $evaluatedExpression, 1);
					}
					foreach ($tagList as $aTag) {
						$tags[] = intval($aTag);
					}
				}
			}
			catch (Exception $e) {
				// Do nothing if expression parsing failed
			}
		}
		return $tags;
	}
}


if (defined('TYPO3_MODE') && $TYPO3_CONF_VARS[TYPO3_MODE]['XCLASS']['ext/tagpackprovider/class.tx_tagpackprovider.php'])	{
	include_once($TYPO3_CONF_VARS[TYPO3_MODE]['XCLASS']['ext/tagpackprovider/class.tx_tagpackprovider.php']);
}
?>

class.tx_tagpackprovider_tca.php
<?php
/***************************************************************
*  Copyright notice
*
*  (c) 2009 Francois Suter (Cobweb) <typo3@cobweb.ch>
*  All rights reserved
*
*  This script is part of the TYPO3 project. The TYPO3 project is
*  free software; you can redistribute it and/or modify
*  it under the terms of the GNU General Public License as published by
*  the Free Software Foundation; either version 2 of the License, or
*  (at your option) any later version.
*
*  The GNU General Public License can be found at
*  http://www.gnu.org/copyleft/gpl.html.
*
*  This script is distributed in the hope that it will be useful,
*  but WITHOUT ANY WARRANTY; without even the implied warranty of
*  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
*  GNU General Public License for more details.
*
*  This copyright notice MUST APPEAR in all copies of the script!
***************************************************************/

/**
 * This class provides a method for filling up the selection of tables in tx_tagpackprovider_selections tables
 *
 * @author		Francois Suter (Cobweb) <typo3@cobweb.ch>
 * @package		TYPO3
 * @subpackage	tx_tagpackprovider
 *
 * $Id$
 */
class tx_tagpackprovider_tca {

	/**
	 * This method provides a list of tables, based on various settings
	 *
	 * @param array $params Properties of the field being modified
	 * @param object $pObj Parent object
	 * @return void
	 */
	function getListOfTables(&$params, &$pObj) {
			// Get the extension configuration
		$configuration = unserialize($GLOBALS['TYPO3_CONF_VARS']['EXT']['extConf']['tagpackprovider']);
			// Get the list of allowed tables from the extension's configuration
			// If empty, get the list of allowed tables from the configuration of tagpack
		if (empty($configuration['taggedTables'])) {
			$tagpackConfiguration = unserialize($GLOBALS['TYPO3_CONF_VARS']['EXT']['extConf']['tagpack']);
			$tablesList = $tagpackConfiguration['taggedTables'];
		} else {
			$tablesList = $configuration['taggedTables'];
		}
			// If a list of tables is defined, use it to fill the items array
		if (!empty($tablesList)) {
			$listOfTables = t3lib_div::trimExplode(',', $tablesList, true);
			foreach ($listOfTables as $table) {
				$label = isset($GLOBALS['TCA'][$table]['ctrl']['label']) ? $GLOBALS['LANG']->sL($GLOBALS['TCA'][$table]['ctrl']['title']) : $table;
				$params['items'][] = array($label, $table);
			}
		}
	}
}

if (defined('TYPO3_MODE') && $TYPO3_CONF_VARS[TYPO3_MODE]['XCLASS']['ext/tagpackprovider/class.tx_tagpackprovider_tca.php']) {
	include_once($TYPO3_CONF_VARS[TYPO3_MODE]['XCLASS']['ext/tagpackprovider/class.tx_tagpackprovider_tca.php']);
}
?>

ext_emconf.php
<?php

########################################################################
# Extension Manager/Repository config file for ext "tagpackprovider".
#
# Auto generated 10-07-2012 10:43
#
# Manual updates:
# Only the data in the array - everything else is removed by next
# writing. "version" and "dependencies" must not be touched!
########################################################################

$EM_CONF[$_EXTKEY] = array(
	'title' => 'Tag Pack-based Data Provider - Tesseract project',
	'description' => 'This Data Provider relies on tags from extension Tag Pack to provide lists of items. More info on http://www.typo3-tesseract.com/',
	'category' => 'services',
	'author' => 'Francois Suter (Cobweb)',
	'author_email' => 'typo3@cobweb.ch',
	'shy' => '',
	'dependencies' => 'tagpack,tesseract,expressions',
	'conflicts' => '',
	'priority' => '',
	'module' => '',
	'state' => 'stable',
	'internal' => '',
	'uploadfolder' => 0,
	'createDirs' => '',
	'modify_tables' => '',
	'clearCacheOnLoad' => 0,
	'lockType' => '',
	'author_company' => '',
	'version' => '1.2.0',
	'constraints' => array(
		'depends' => array(
			'tagpack' => '',
			'typo3' => '4.5.0-4.7.99',
			'tesseract' => '1.0.0-0.0.0',
			'expressions' => '',
		),
		'conflicts' => array(
		),
		'suggests' => array(
		),
	),
	'_md5_values_when_last_written' => 'a:17:{s:9:"ChangeLog";s:4:"d623";s:28:"class.tx_tagpackprovider.php";s:4:"c151";s:32:"class.tx_tagpackprovider_tca.php";s:4:"ec88";s:16:"ext_autoload.php";s:4:"d40e";s:21:"ext_conf_template.txt";s:4:"6374";s:12:"ext_icon.gif";s:4:"f4d1";s:17:"ext_localconf.php";s:4:"4cba";s:14:"ext_tables.php";s:4:"e9a1";s:14:"ext_tables.sql";s:4:"5796";s:45:"locallang_csh_txtagpackproviderselections.xml";s:4:"e45c";s:16:"locallang_db.xml";s:4:"7a77";s:10:"README.txt";s:4:"a890";s:7:"tca.php";s:4:"4ded";s:14:"doc/manual.pdf";s:4:"6f68";s:14:"doc/manual.sxw";s:4:"5451";s:34:"res/add_tagpackprovider_wizard.gif";s:4:"032f";s:42:"res/icon_tx_tagpackprovider_selections.gif";s:4:"f4d1";}',
	'suggests' => array(
	),
);

?>